File: backend/routers/annotations.py
"""Annotations API routes. Thin layer: dependencies and controller delegation."""

from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from deps import get_db
from auth import get_current_active_user
from models.user import User
from schemas.annotation import (
    AnnotationCreate,
    AnnotationUpdate,
    AnnotationResponse,
    ValidatePositionsRequest,
    BulkCreateAnnotationsRequest,
    BulkDeleteByCriteriaRequest,
)

from controllers import annotations as annotations_controller

router = APIRouter(prefix="/annotations", tags=["Annotations"])


@router.get("/", response_model=List[AnnotationResponse])
def read_annotations(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    text_id: Optional[int] = Query(None),
    annotator_id: Optional[int] = Query(None),
    annotation_type: Optional[str] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get annotations list with optional filtering."""
    return annotations_controller.read_annotations(
        db,
        current_user,
        skip=skip,
        limit=limit,
        text_id=text_id,
        annotator_id=annotator_id,
        annotation_type=annotation_type,
    )


@router.post("/", response_model=AnnotationResponse, status_code=201)
def create_annotation(
    annotation_in: AnnotationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Create new annotation."""
    return annotations_controller.create_annotation(
        db, current_user, annotation_in
    )


@router.get("/my-annotations", response_model=List[AnnotationResponse])
def read_my_annotations(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get current user's annotations."""
    return annotations_controller.read_my_annotations(
        db, current_user, skip=skip, limit=limit
    )


@router.get("/text/{text_id}", response_model=List[AnnotationResponse])
def read_annotations_by_text(
    text_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get all annotations for a specific text."""
    return annotations_controller.read_annotations_by_text(
        db, current_user, text_id
    )


@router.get("/type/{annotation_type}", response_model=List[AnnotationResponse])
def read_annotations_by_type(
    annotation_type: str,
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get annotations by type."""
    return annotations_controller.read_annotations_by_type(
        db, current_user, annotation_type, skip=skip, limit=limit
    )


@router.get("/stats")
def get_annotation_stats(
    text_id: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get annotation statistics."""
    return annotations_controller.get_annotation_stats(
        db, current_user, text_id=text_id
    )


@router.get("/{annotation_id}", response_model=AnnotationResponse)
def read_annotation(
    annotation_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Get annotation by ID."""
    return annotations_controller.read_annotation(
        db, current_user, annotation_id
    )


@router.put("/{annotation_id}", response_model=AnnotationResponse)
def update_annotation(
    annotation_id: int,
    annotation_in: AnnotationUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Update annotation."""
    return annotations_controller.update_annotation(
        db, current_user, annotation_id, annotation_in
    )


@router.delete("/{annotation_id}", status_code=204)
def delete_annotation(
    annotation_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Delete annotation."""
    annotations_controller.delete_annotation(db, current_user, annotation_id)


@router.delete("/text/{text_id}/my-annotations")
def delete_my_annotations_for_text(
    text_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Delete all of the current user's annotations for a specific text."""
    return annotations_controller.delete_my_annotations_for_text(
        db, current_user, text_id
    )


@router.post("/validate-positions")
def validate_annotation_positions(
    body: ValidatePositionsRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Validate annotation positions before creating/updating."""
    return annotations_controller.validate_annotation_positions(
        db,
        current_user,
        body.text_id,
        body.start_position,
        body.end_position,
    )


@router.post("/bulk-create")
def bulk_create_annotations(
    body: BulkCreateAnnotationsRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Create annotations at all given spans in one request (apply to all)."""
    return annotations_controller.bulk_create_annotations(db, current_user, body)


@router.post("/bulk-delete")
def bulk_delete_annotations(
    body: BulkDeleteByCriteriaRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user),
):
    """Delete all annotations matching type/label/selected_text for the text (delete from all)."""
    return annotations_controller.bulk_delete_annotations(db, current_user, body)

File: frontend/src/api/annotations.ts
import { apiClient } from "./utils";
import type {
  AnnotationResponse,
  AnnotationCreate,
  AnnotationUpdate,
  AnnotationFilters,
  AnnotationStats,
  ValidatePositionsRequest,
  ValidatePositionsResponse,
  DeleteMyAnnotationsResponse,
  BulkCreateAnnotationsRequest,
  BulkCreateAnnotationsResponse,
  BulkDeleteByCriteriaRequest,
  BulkDeleteAnnotationsResponse,
} from "./types";

export interface CustomAnnotationLabelResponse {
  label: string;
  annotation_type_id?: string;
  annotation_type_name?: string;
  usage_count: number;
  user_count: number;
  text_count: number;
  first_seen_at: string;
  last_seen_at: string;
}

// Annotations API client
export const annotationsApi = {
  // Get all annotations with optional filtering
  getAnnotations: async (
    filters: AnnotationFilters = {}
  ): Promise<AnnotationResponse[]> => {
    return apiClient.get<AnnotationResponse[]>("/annotations", filters);
  },

  // Get single annotation by ID
  getAnnotation: async (id: number): Promise<AnnotationResponse> => {
    return apiClient.get<AnnotationResponse>(`/annotations/${id}`);
  },

  // Create new annotation
  createAnnotation: async (
    data: AnnotationCreate
  ): Promise<AnnotationResponse> => {
    return apiClient.post<AnnotationResponse>("/annotations", data);
  },

  // Update annotation
  updateAnnotation: async (
    id: number,
    data: AnnotationUpdate
  ): Promise<AnnotationResponse> => {
    return apiClient.put<AnnotationResponse>(`/annotations/${id}`, data);
  },

  // Delete annotation
  deleteAnnotation: async (id: number): Promise<void> => {
    return apiClient.delete<void>(`/annotations/${id}`);
  },

  // Delete my annotations for a text
  deleteMyAnnotationsForText: async (textId: number): Promise<DeleteMyAnnotationsResponse> => {
    return apiClient.delete<DeleteMyAnnotationsResponse>(`/annotations/text/${textId}/my-annotations`);
  },
  // Get current user's annotations
  getMyAnnotations: async (
    filters: { skip?: number; limit?: number } = {}
  ): Promise<AnnotationResponse[]> => {
    return apiClient.get<AnnotationResponse[]>(
      "/annotations/my-annotations",
      filters
    );
  },

  // Get annotations by text ID
  getAnnotationsByText: async (
    textId: number
  ): Promise<AnnotationResponse[]> => {
    return apiClient.get<AnnotationResponse[]>(`/annotations/text/${textId}`);
  },

  // Get annotations by type
  getAnnotationsByType: async (
    annotationType: string,
    filters: { skip?: number; limit?: number } = {}
  ): Promise<AnnotationResponse[]> => {
    return apiClient.get<AnnotationResponse[]>(
      `/annotations/type/${annotationType}`,
      filters
    );
  },

  // Get annotation statistics
  getAnnotationStats: async (textId?: number): Promise<AnnotationStats> => {
    const params = textId ? { text_id: textId } : {};
    return apiClient.get<AnnotationStats>("/annotations/stats", params);
  },

  // Get unique custom annotation labels used by users but missing from the canonical annotation list
  getCustomAnnotationLabels: async (): Promise<CustomAnnotationLabelResponse[]> => {
    return apiClient.get<CustomAnnotationLabelResponse[]>("/annotations/custom-labels");
  },

  // Validate annotation positions
  validatePositions: async (
    data: ValidatePositionsRequest
  ): Promise<ValidatePositionsResponse> => {
    return apiClient.post<ValidatePositionsResponse>(
      "/annotations/validate-positions",
      data
    );
  },

  // Bulk create (apply to all)
  bulkCreateAnnotations: async (
    data: BulkCreateAnnotationsRequest
  ): Promise<BulkCreateAnnotationsResponse> => {
    return apiClient.post<BulkCreateAnnotationsResponse>(
      "/annotations/bulk-create",
      data
    );
  },

  // Bulk delete (delete from all)
  bulkDeleteAnnotations: async (
    data: BulkDeleteByCriteriaRequest
  ): Promise<BulkDeleteAnnotationsResponse> => {
    return apiClient.post<BulkDeleteAnnotationsResponse>(
      "/annotations/bulk-delete",
      data
    );
  },
};
